twitterscraper/query_js.py

```diff
@@ -0,0 +1,197 @@
+
+    try:
+        data = defaultdict(dict)
+        already_idxs = set()
+
+        # page down, recording the results, until there isn't anything new or limit has been breached
+        driver.get(url)
+        retries = 20
+        while retries > 0 and len(data['tweets']) < limit:
+
+            # relevant requests have completely responses, json in their path (but not guide.json), and a globalObjects key
+            relevant_request_idxs = set([
+                i for i, r in enumerate(driver.requests)
+                if 'json' in r.path and 'guide.json' not in r.path and
+                r.response is not None and isinstance(r.response.body, dict) and
+                'globalObjects' in r.response.body and i not in already_idxs
+            ])
+            already_idxs |= relevant_request_idxs
+
+            if not relevant_request_idxs:
+                time.sleep(0.2)
+                retries -= 1
+                continue
+
+            # if no relevant requests, or latest relevant request isn't done loading, wait then check again
+            latest_tweets = driver.requests[max(relevant_request_idxs)].response.body['globalObjects']['tweets']
+            if len(relevant_request_idxs) == 0 or not latest_tweets:
+                time.sleep(0.2)
+                retries -= 1
+                continue
+
+            # scroll down
+            actions = ActionChains(driver)
+            for _ in range(100):
+                actions.send_keys(Keys.PAGE_DOWN)
+            actions.perform()
+
+            # record relevant responses
+            for idx in relevant_request_idxs:
+                driver.requests[idx]
+                for key, value in driver.requests[idx].response.body['globalObjects'].items():
+                    data[key].update(value)
+
+            # reset retries
+            retries = 20
+
+        return data
+
+    except Exception as e:
+        logger.exception('Exception {} while requesting "{}"'.format(
+            e, url))
+    finally:
+        driver.quit()
+
+    if retry > 0:
+        logger.debug('Retrying... (Attempts left: {})'.format(retry))
+        return query_single_page(url, retry - 1)
+    logger.error('Giving up.')
+    return defaultdict(dict)
+
+
+def get_query_data(queries, limit=None, begindate=None, enddate=None, poolsize=None, lang=''):
+    begindate = begindate or dt.date(2006, 3, 21)
+    enddate = enddate or dt.date.today()
+    poolsize = poolsize or 5
+
+    num_days = (enddate - begindate).days
+
+    if(num_days < 0):
+        sys.exit('Begin date must occur before end date.')
+
+    if poolsize > num_days:
+        # Since we are assigning each pool a range of dates to query,
+        # the number of pools should not exceed the number of dates.
+        poolsize = num_days
+    # query one day at a time so driver doesn't use too much memory
+    dateranges = list(reversed([begindate + dt.timedelta(days=elem) for elem in linspace(0, num_days, num_days)]))
```

Comment:
linspace(0, num_days, num_days) **always** creates num_days amount of separate queries instead of poolsize+1 number of separate queries . 
Please have a look at query.py to see how it was originally implemented. 
`dateranges = [begindate + dt.timedelta(days=elem) for elem in linspace(0, no_days, poolsize+1)]`

Reply:
This behavior is different because

1) the pool size is limited so it doesn't matter how many items we have in our list
2) we want more granularity in case loading a history fails. If we have 1000 days, and twitterscraper fails to load a day 10% of the time, we'll run twitterscraper for a single day 1,100 times on average. If our interval is one month instead of one day, we'll succeed retrieving a month 4% (90%^30) of the time, resulting in 20,000 work units instead of 1,100.
3) We save precious memory by only going one day at the time. Emulated browsers store increasingly more state as we scroll further.

Reply:
Sounds reasonable. 
Do you think creating a separate query for each day also has the same benefits (indicated under 2) for regular old query.py?

Reply:
I don't think it's necessary for old `query.py` because it already shards and retries on a granular level, and there isn't the risk of needing to "restart" from the beginning of the day. Perhaps it would be nice for the two APIs to be consistent, however.